platform/linux-generic/odp_schedule_scalable.c

```diff
@@ -958,9 +961,9 @@ static int _schedule(odp_queue_t *from, odp_event_t ev[], int num_evts)
 		}
 		if (is_pktin(elem)) {
 			/* Pktio ingress queue */
-			if (first == NULL)
-				first = elem;
-			else if (elem == first) /* Wrapped around */
+			if (elem->loop_check[cpu_id] != ts->loop_cnt)
```

Comment:
One last thing. Have you tried adding an extra if-clause here to check if the element was received from a lower priority pktio scheduling queue? If I've understood the `poll_pktin()` code correctly, the dead loop issue only affects the low priority queues, so this way the overhead would be minimized for normal priority queues.

E.g.
```
if (elem->schedq != schedq) {
	if (elem->loop_check[cpu_id] != ts->loop_cnt)
		elem->loop_check[cpu_id] = ts->loop_cnt;
	...
}
```

Reply:
i haven't  tried adding extra if-clause.     
as i understand, there are too possible  moving actions for elem in poll_pktin(): 
from low to high: schedq_push(elem->schedq, elem);   //deadloop risk for  lower priority pktio scheduling queue
from high to low : schedq += elem->xoffset;schedq_push(schedq, elem); //deadloop risk for  higher priority pktio scheduling queue . 
  
  i don't understand:   if (first == NULL), is still needed.

Reply:
In case of higher priority queue there shouldn't be a dead loop risk, since in the worst case the scheduling queue will eventually be empty (all pktin elems have been moved to the lower prio).

The `if (first == NULL)` check is needed because ordered queues are returned back to the same priority scheduling queue if the reorder context is not available.

Reply:
U are right, higher queue dont have the risk.    how to judge high ?  if we add a judgement for queue index, it also cost some cpu cycle.

Reply:
The original sched queue is saved in `elem->schedq`, so that can be compared to the current `schedq` -> `if (elem->schedq != schedq) ` -> lower prio sched queue.

I would assume that the cost of adding the comparison here is quite minimal (`elem->schedq` is read anyway inside `poll_pktin()`) and it would be worth it to avoid having to write to every element in the high/normal prio queue.

I'm currently working from home and have no way of testing this. Can you test if this optimization works or should we go first with the current fix and do the possible optimiztion in a separate PR?

Reply:
hi  matiaelo
   
if you agree, i will commit it . i think we also need not to use first . 

		if (is_pktin(elem)) {
			/* Pktio ingress queue */
			if (elem->schedq != schedq) {
				if (elem->loop_check[cpu_id] != ts->loop_cnt)
					elem->loop_check[cpu_id] = ts->loop_cnt;
				else /* Wrapped around */
					continue; /* Go to next schedq */
			}

			if (odp_unlikely(!schedq_cond_pop(schedq, elem)))
				goto restart_same;

Reply:
This looks good 👍

While you do this, please also add a short description (a sentence or two) of the problem/solution and a Signed-off-by tag to the commit log.

Reply:
Signed-off-by: Matias Elo <matias.elo@nokia.com>
Reviewed-by: Matias Elo <matias.elo@nokia.com>
 is it ok  ?

Reply:
Signed-off-by you , Reviewed-by me =)